Mathlib/Topology/Algebra/FilterBasisNew.lean

```diff
@@ -0,0 +1,448 @@
+one satisfying `Filter.IsGroupBasis`. -/
+@[to_additive "Assume the group `G` is given a topology which is invariant by translations,
+which we express as `ContinuousConstVAdd G G`.
+To show that it is a group topology, it suffices so exhibit a basis of neighborhoods of
+zero satisfying `Filter.IsAddGroupBasis`."]
+lemma topologicalGroup_of_hasBasis [TopologicalSpace G] [ContinuousConstSMul G G]
+    (hB' : (𝓝 1).HasBasis p s) : TopologicalGroup G := by
+  apply TopologicalGroup.of_nhds_one
+  · refine hB'.prod_self.tendsto_iff hB' |>.mpr fun i hi ↦ (hB.mul hi).imp
+      fun j ⟨hj, hji⟩ ↦ ⟨hj, ?_⟩
+    simpa [← image2_mul, forall_mem_comm] using hji
+  · exact hB'.tendsto_iff hB' |>.mpr fun i hi ↦ (hB.inv hi).imp fun j ↦ id
+  · intro x₀
+    simp_rw [← smul_eq_mul, ← Homeomorph.smul_apply x₀, (Homeomorph.smul x₀).map_nhds_eq,
+      Homeomorph.smul_apply x₀, smul_eq_mul, mul_one]
+  · exact fun x₀ ↦ hB'.tendsto_iff hB' |>.mpr fun i hi ↦ (hB.conj x₀ hi).imp fun j ↦ id
+
+/-!
+### Constructing a group topology from `Filter.IsGroupBasis`
+-/
+
+/-- The topological space structure coming from a group filter basis. -/
+@[to_additive "The topological space structure coming from an additive group filter basis."]
+def topology : TopologicalSpace G :=
+  TopologicalSpace.mkOfNhds (· • hB.filter)
+
+@[to_additive]
+theorem nhds_eq {x₀ : G} : @nhds G hB.topology x₀ = x₀ • hB.filter := by
+  apply TopologicalSpace.nhds_mkOfNhds_of_hasBasis (fun x ↦ hB.hasBasis.map (x * ·))
+  · intro a i hi
+    exact ⟨1, hB.one hi, mul_one a⟩
+  · intro a i hi
+    rcases hB.mul hi with ⟨j, hj, hji⟩
+    filter_upwards [hB.hasBasis.map _ |>.mem_of_mem hj]
+    rintro _ ⟨x, hx, rfl⟩
+    calc
+      (a * x) • (s j) ∈ (a * x) • hB.filter := smul_set_mem_smul_filter (hB.hasBasis.mem_of_mem hj)
+      _ = a • x • (s j) := smul_smul .. |>.symm
+      _ ⊆ a • (s j * s j) := smul_set_mono <| smul_set_subset_smul hx
+      _ ⊆ a • (s i) := smul_set_mono hji
+
+@[to_additive]
+theorem nhds_one_eq :
+    @nhds G hB.topology (1 : G) = hB.filter := by
+  rw [hB.nhds_eq, one_smul]
+
+@[to_additive]
+theorem nhds_hasBasis (x₀ : G) :
+    HasBasis (@nhds G hB.topology x₀) p (fun i ↦ x₀ • (s i)) := by
+  rw [hB.nhds_eq]
+  apply hB.hasBasis.map _
+
+@[to_additive]
+theorem nhds_one_hasBasis :
+    HasBasis (@nhds G hB.topology 1) p s := by
+  rw [hB.nhds_one_eq]
+  exact hB.hasBasis
+
+@[to_additive]
+theorem mem_nhds_one {i} (hi : p i) :
+    s i ∈ @nhds G hB.topology 1 :=
+  hB.nhds_one_hasBasis.mem_of_mem hi
+
+-- See note [lower instance priority]
+/-- If a group is endowed with the topological structure coming from a group filter basis, then
+(by definition) this topology is invariant by translations. -/
+@[to_additive "If a group is endowed with the topological structure coming from a group filter
+basis, then (by definition) this topology is invariant by translations."]
+instance (priority := 100) continuousConstSMul :
```

Comment:
I have no way to assess the relevance of this priority level. Why 100, why not 200 or 50?

Reply:
Most instances of `ContinuousConstSMul` have the default priority of 1000, so anything lower will just be considered later. Here, that's because we don't want Lean to spend ages trying to unify the topology on G with a topology of this exact form. Instead, any user defining a topology in this way should repeat the instance (with usual priority) in that special setting. As for 100/200/50, the truth is it probably doesn't matter because there's not that many instances of `ContinuousConstSMul` to consider in the hierarchy. What matters is that it's below 1000.